monai/data/image_writer.py

```diff
@@ -0,0 +1,725 @@
+        shape `(64, 64, 8)` or `(64, 64, 8, 1)` will be considered as a
+        single-channel 3D image. The ``convert_to_channel_last`` method can be
+        used to convert the data to the format described here.
+
+        Note that the shape of the resampled ``data_array`` may subject to some
+        rounding errors. For example, resampling a 20x20 pixel image from pixel
+        size (1.5, 1.5)-mm to (3.0, 3.0)-mm space will return a 10x10-pixel
+        image. However, resampling a 20x20-pixel image from pixel size (2.0,
+        2.0)-mm to (3.0, 3.0)-mm space will output a 14x14-pixel image, where
+        the image shape is rounded from 13.333x13.333 pixels. In this case
+        ``output_spatial_shape`` could be specified so that this function
+        writes image data to a designated shape.
+
+        Args:
+            data_array: input data array to be converted.
+            affine: the current affine of ``data_array``. Defaults to identity
+            target_affine: the designated affine of ``data_array``.
+                The actual output affine might be different from this value due to precision changes.
+            output_spatial_shape: spatial shape of the output image.
+                This option is used when resampling is needed.
+            mode: available options are {``"bilinear"``, ``"nearest"``, ``"bicubic"``}.
+                This option is used when resampling is needed.
+                Interpolation mode to calculate output values. Defaults to ``"bilinear"``.
+                See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+            padding_mode: available options are {``"zeros"``, ``"border"``, ``"reflection"``}.
+                This option is used when resampling is needed.
+                Padding mode for outside grid values. Defaults to ``"border"``.
+                See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+            align_corners: boolean option of ``grid_sample`` to handle the corner convention.
+                See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+            dtype: data type for resampling computation. Defaults to
+                ``np.float64`` for best precision. If ``None``, use the data type of input data.
+        """
+        data: np.ndarray
+        data, *_ = convert_data_type(data_array, np.ndarray)  # type: ignore
+
+        sr = min(data.ndim, 3)
+        if affine is None:
+            affine = np.eye(4, dtype=np.float64)
+        affine = to_affine_nd(sr, affine)  # type: ignore
+        target_affine = to_affine_nd(sr, target_affine) if target_affine is not None else affine
+
+        if np.allclose(affine, target_affine, atol=AFFINE_TOL):
+            # no affine changes, return (data, affine)
+            return data, ensure_mat44(target_affine)
+
+        # resolve orientation
+        start_ornt = nib.orientations.io_orientation(affine)
```

Comment:
It might make sense to put this code into a separate utility function if the same sort of operation is needed elsewhere:

```python
def convert_nifti_transform(init_affine, target_affine, data_shape):
    """Return an NIfTI-resolved affine representing the transform from `init_affine` space to `target_affine` space."""
    start_ornt = nib.orientations.io_orientation(init_affine)
    target_ornt = nib.orientations.io_orientation(target_affine)
    ornt_transform = nib.orientations.ornt_transform(start_ornt, target_ornt)
    _affine = affine @ nib.orientations.inv_ornt_aff(ornt_transform, data_shape)
    return _affine
```

(I think the docstring is right but do check)

Reply:
thanks, addressed by https://github.com/Project-MONAI/MONAI/pull/3674/commits/e189fbfc155670a71a4a198e64a1e0fca05d7a5e.